src/wallet/rpcwallet.cpp

```diff
@@ -4073,26 +4085,31 @@ static RPCHelpMan send()
                 UniValueType(), // ARR or OBJ, checked later
                 UniValue::VNUM,
                 UniValue::VSTR,
-                UniValue::VOBJ
+                UniValue::VNUM,
+                UniValue::VOBJ,
                 }, true
             );
 
             std::shared_ptr<CWallet> const wallet = GetWalletForJSONRPCRequest(request);
             if (!wallet) return NullUniValue;
             CWallet* const pwallet = wallet.get();
 
-            UniValue options{request.params[3].isNull() ? UniValue::VOBJ : request.params[3]};
-            if (options.exists("feeRate") || options.exists("fee_rate") || options.exists("estimate_mode") || options.exists("conf_target")) {
-                if (!request.params[1].isNull() || !request.params[2].isNull()) {
-                    throw JSONRPCError(RPC_INVALID_PARAMETER, "Use either conf_target and estimate_mode or the options dictionary to control fee rate");
+            UniValue options{request.params[4].isNull() ? UniValue::VOBJ : request.params[4]};
+            if (options.exists("conf_target") || options.exists("estimate_mode") || options.exists("fee_rate")) {
+                if (!request.params[1].isNull() || !request.params[2].isNull() || !request.params[3].isNull()) {
```

Comment:
Maybe I'm missing something, but I had trouble figuring out from the local code what the parameter set of this particular call is here. If I understand correctly, their types are being defined in 4086-4089. Maybe the parameter names could be added as comments there or here to remind the reader what is being tested to be null here in 4099.

Reply:
I was trying to figure out what exactly is checked here that prevents e.g. both `conf_target` and `fee_rate` to be set here, if `params[1]…[3]` all need to be null here.

Reply:
Add these comments:
```diff
             RPCTypeCheck(request.params, {
-                UniValueType(), // ARR or OBJ, checked later
-                UniValue::VNUM,
-                UniValue::VSTR,
-                UniValue::VNUM,
-                UniValue::VOBJ,
+                UniValueType(), // outputs (ARR or OBJ, checked later)
+                UniValue::VNUM, // conf_target
+                UniValue::VSTR, // estimate_mode
+                UniValue::VNUM, // fee_rate
+                UniValue::VOBJ, // options
                 }, true
             );
```
If any of these options are passed, then none of the same ones should be passed as arguments. I suppose the logic or error message could be refined further but think I'll punt on that for now. This RPC is still labeled as experimental.

Reply:
> I was trying to figure out what exactly is checked here that prevents e.g. both `conf_target` and `fee_rate` to be set here, if `params[1]…[3]` all need to be null here.

Hm wait, the new error message should be changed. It used to be, "Use either conf_target and estimate_mode or the options dictionary to control fee rate"... will update.

Reply:
Added these comments:
```diff
             RPCTypeCheck(request.params, {
-                UniValueType(), // ARR or OBJ, checked later
-                UniValue::VNUM,
-                UniValue::VSTR,
-                UniValue::VNUM,
-                UniValue::VOBJ,
+                UniValueType(), // outputs (ARR or OBJ, checked later)
+                UniValue::VNUM, // conf_target
+                UniValue::VSTR, // estimate_mode
+                UniValue::VNUM, // fee_rate
+                UniValue::VOBJ, // options
                 }, true
             );
```
> I was trying to figure out what exactly is checked here that prevents e.g. both `conf_target` and `fee_rate` to be set here, if `params[1]…[3]` all need to be null here.

You're right. Updated.
```diff
             UniValue options{request.params[4].isNull() ? UniValue::VOBJ : request.params[4]};
-            if (options.exists("conf_target") || options.exists("estimate_mode") || options.exists("fee_rate")) {
-                if (!request.params[1].isNull() || !request.params[2].isNull() || !request.params[3].isNull()) {
-                    throw JSONRPCError(RPC_INVALID_PARAMETER, "Use either conf_target and estimate_mode for fee estimation, or fee_rate to specify an explicit fee rate.");
+            if (options.exists("conf_target") || options.exists("estimate_mode")) {
+                if (!request.params[1].isNull() || !request.params[2].isNull()) {
+                    throw JSONRPCError(RPC_INVALID_PARAMETER, "Pass conf_target and estimate_mode either as arguments or in the options object, but not both");
                 }
             } else {
                 options.pushKV("conf_target", request.params[1]);
                 options.pushKV("estimate_mode", request.params[2]);
+            }
+            if (options.exists("fee_rate")) {
+                if (!request.params[3].isNull()) {
+                    throw JSONRPCError(RPC_INVALID_PARAMETER, "Pass fee_rate either as an argument or in the options object, but not both");
+                }
+            } else {
                 options.pushKV("fee_rate", request.params[3]);
             }
```
This hopefully maintains the original goal of the code that either the argument or the option should be passed, but not both. The check if conf_target or estimate_mode are passed at the same time as fee_rate happens in `SetFeeEstimateMode()`.

Reply:
Updated `wallet_send` tests.
```diff
+++ b/test/functional/wallet_send.py
@@ -234,7 +234,7 @@ class WalletSendTest(BitcoinTestFramework):
         for mode in ["unset", "economical", "conservative"]:
             self.test_send(from_wallet=w0, to_wallet=w1, amount=1, arg_conf_target=1, arg_estimate_mode="economical",
                 conf_target=1, estimate_mode=mode, add_to_wallet=False,
-                expect_error=(-8, "Use either conf_target and estimate_mode for fee estimation, or fee_rate to specify an explicit fee rate."))
+                expect_error=(-8, "Pass conf_target and estimate_mode either as arguments or in the options object, but not both"))
 
         self.log.info("Create PSBT from watch-only wallet w3, sign with w2...")
         res = self.test_send(from_wallet=w3, to_wallet=w1, amount=1)
@@ -278,6 +278,10 @@ class WalletSendTest(BitcoinTestFramework):
         fee = self.nodes[1].decodepsbt(res["psbt"])["fee"]
         assert_fee_amount(fee, Decimal(len(res["hex"]) / 2), Decimal("0.00003"))
 
+        # Test that passing fee_rate as both an argument and an option raises.
+        self.test_send(from_wallet=w0, to_wallet=w1, amount=1, arg_fee_rate=1, fee_rate=1, add_to_wallet=False,
+                       expect_error=(-8, "Pass fee_rate either as an argument or in the options object, but not both"))
```

Reply:
Ooooh, so this was checking that not both the options and arguments were being passed. I guess I hadn't fully grokked that. I guess that would have been fine then, if any of the three was set on one, it would be incorrect for the any of the others to be set in the other parameter set. It actually is a bit more lenient locally then, but since e.g. passing both estimate_mode and fee_rate would throw in other places this is not a problem.

Reply:
Thanks to your feedback, the error messages seem clearer now.